[AppBar] Add a shouldSetNavigationBarHiddenHideAppBar behavior flag.

jverkoey opened the pull request:
This new flag allows view controllers to control the visibility of their app bar via the standard UINavigationController setNavigationBarHidden: APIs. When enabled, calls to these APIs will result in the AppBar's visibility being adjusted.

Closes https://github.com/material-components/material-components-ios/issues/5185

material-automation commented on components/AppBar/src/MDCAppBarNavigationController.m:
```diff
@@ -64,8 +68,10 @@ - (instancetype)initWithRootViewController:(UIViewController *)rootViewControlle
-  [self setNavigationBarHidden:YES animated:NO];
+  _childNavigationBarsHidden = NO;
+
+  // We always want the UIKit navigation bar to be hidden; to do so we must invoke the super implementation.
```
clang-format suggested the following change:


```
-  // We always want the UIKit navigation bar to be hidden; to do so we must invoke the super implementation.
+  // We always want the UIKit navigation bar to be hidden; to do so we must invoke the super
+  // implementation.
```

<!-- identifier: clang-format -->
<!-- File: components/AppBar/src/MDCAppBarNavigationController.m -->

material-automation commented on components/AppBar/src/MDCAppBarNavigationController.m:
```diff
@@ -102,19 +108,32 @@ - (void)setViewControllers:(NSArray<UIViewController *> *)viewControllers animat
+  _childNavigationBarsHidden = navigationBarHidden;
+
+  for (UIViewController *viewController in self.viewControllers) {
+    MDCAppBarViewController *appBarViewController = [self appBarViewControllerForViewController:viewController];
```
clang-format suggested the following change:


```
-    MDCAppBarViewController *appBarViewController = [self appBarViewControllerForViewController:viewController];
+    MDCAppBarViewController *appBarViewController =
+        [self appBarViewControllerForViewController:viewController];
```

<!-- identifier: clang-format -->
<!-- File: components/AppBar/src/MDCAppBarNavigationController.m -->

material-automation commented on components/FlexibleHeader/src/MDCFlexibleHeaderView+ShiftBehavior.h:
```diff
@@ -38,6 +38,13 @@ typedef NS_ENUM(NSInteger, MDCFlexibleHeaderShiftBehavior) {
+  /**
+   Similar to MDCFlexibleHeaderShiftBehaviorEnabled, but cannot be interactively hidden or shown and
+   the flexible header can only be shifted on or off-screen using the @c shiftHeaderOnScreenAnimated:
+   and @c shiftHeaderOffScreenAnimated APIs.
```
clang-format suggested the following change:


```
-   the flexible header can only be shifted on or off-screen using the @c shiftHeaderOnScreenAnimated:
-   and @c shiftHeaderOffScreenAnimated APIs.
+   the flexible header can only be shifted on or off-screen using the @c
+   shiftHeaderOnScreenAnimated: and @c shiftHeaderOffScreenAnimated APIs.
```

<!-- identifier: clang-format -->
<!-- File: components/FlexibleHeader/src/MDCFlexibleHeaderView+ShiftBehavior.h -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

material-automation commented:
bazel detected changes to the following targets:

```
//components/AppBar:snapshot_tests
//components/AppBar:snapshot_tests_test_bundle
//components/AppBar:unit_tests
//components/AppBar:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/AppBar:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/AppBar:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/AppBar:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/AppBar:unit_tests_environment
//components/AppBar:unit_tests_environment_test_bundle
//components/Tabs:snapshot_tests
//components/Tabs:snapshot_tests_test_bundle
```

<!-- identifier: affected-targets -->

material-automation commented:
## API diff detected the following changes

### AppBar

#### MDCAppBarNavigationController

*new* property: `shouldSetNavigationBarHiddenHideAppBar` in `MDCAppBarNavigationController`

*new* property: `navigationBarHidden` in `MDCAppBarNavigationController`

<!-- identifier: apidiff -->

material-automation commented on components/FlexibleHeader/src/MDCFlexibleHeaderView.m:
```diff
@@ -1044,7 +1044,8 @@ - (void)fhv_updateLayout {
 
-  if (_shiftAccumulatorLastContentOffsetIsValid) {
+  if (_shiftAccumulatorLastContentOffsetIsValid
+      && _shiftBehavior != MDCFlexibleHeaderShiftBehaviorHideable) {
```
clang-format suggested the following change:


```
-  if (_shiftAccumulatorLastContentOffsetIsValid
-      && _shiftBehavior != MDCFlexibleHeaderShiftBehaviorHideable) {
+  if (_shiftAccumulatorLastContentOffsetIsValid &&
+      _shiftBehavior != MDCFlexibleHeaderShiftBehaviorHideable) {
```

<!-- identifier: clang-format -->
<!-- File: components/FlexibleHeader/src/MDCFlexibleHeaderView.m -->

material-automation commented on components/FlexibleHeader/src/MDCFlexibleHeaderView.m:
```diff
@@ -1686,7 +1687,8 @@ - (void)setStatusBarHintCanOverlapHeader:(BOOL)statusBarHintCanOverlapHeader {
   NSAssert((self.observesTrackingScrollViewScrollEvents &&
-            shiftBehavior == MDCFlexibleHeaderShiftBehaviorDisabled) ||
+            (shiftBehavior == MDCFlexibleHeaderShiftBehaviorDisabled
+             || shiftBehavior == MDCFlexibleHeaderShiftBehaviorHideable)) ||
```
clang-format suggested the following change:


```
-            (shiftBehavior == MDCFlexibleHeaderShiftBehaviorDisabled
-             || shiftBehavior == MDCFlexibleHeaderShiftBehaviorHideable)) ||
+            (shiftBehavior == MDCFlexibleHeaderShiftBehaviorDisabled ||
+             shiftBehavior == MDCFlexibleHeaderShiftBehaviorHideable)) ||
```

<!-- identifier: clang-format -->
<!-- File: components/FlexibleHeader/src/MDCFlexibleHeaderView.m -->

material-automation commented on components/AppBar/src/MDCAppBarNavigationController.m:
```diff
@@ -102,23 +109,77 @@ - (void)setViewControllers:(NSArray<UIViewController *> *)viewControllers animat
+                             animated:(BOOL)animated
+                    forViewController:(UIViewController *)viewController {
+  MDCAppBarViewController *appBarViewController =
+  [self appBarViewControllerForViewController:viewController];
```
clang-format suggested the following change:


```
-  [self appBarViewControllerForViewController:viewController];
+      [self appBarViewControllerForViewController:viewController];
```

<!-- identifier: clang-format -->
<!-- File: components/AppBar/src/MDCAppBarNavigationController.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

material-automation commented on components/FlexibleHeader/src/MDCFlexibleHeaderView.m:
```diff
@@ -1333,9 +1336,10 @@ - (void)fhv_stopObservingContentOffset {
-           @"Please set shiftBehavior to disabled prior to enabling this property.");
+  NSAssert(!observesTrackingScrollViewScrollEvents ||
+           (self.shiftBehavior == MDCFlexibleHeaderShiftBehaviorDisabled
+            || self.shiftBehavior == MDCFlexibleHeaderShiftBehaviorHideable),
```
clang-format suggested the following change:


```
-           (self.shiftBehavior == MDCFlexibleHeaderShiftBehaviorDisabled
-            || self.shiftBehavior == MDCFlexibleHeaderShiftBehaviorHideable),
+               (self.shiftBehavior == MDCFlexibleHeaderShiftBehaviorDisabled ||
+                self.shiftBehavior == MDCFlexibleHeaderShiftBehaviorHideable),
```

<!-- identifier: clang-format -->
<!-- File: components/FlexibleHeader/src/MDCFlexibleHeaderView.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

material-automation commented on components/FlexibleHeader/src/MDCFlexibleHeaderView.m:
```diff
@@ -521,7 +521,8 @@ - (void)flexibleHeaderSafeAreaTopSafeAreaInsetDidChange:(MDCFlexibleHeaderTopSaf
   return ([self fhv_canShiftOffscreen] &&
-          _shiftBehavior == MDCFlexibleHeaderShiftBehaviorEnabledWithStatusBar &&
+          (_shiftBehavior == MDCFlexibleHeaderShiftBehaviorEnabledWithStatusBar
+           || _shiftBehavior == MDCFlexibleHeaderShiftBehaviorHideable) &&
```
clang-format suggested the following change:


```
-          (_shiftBehavior == MDCFlexibleHeaderShiftBehaviorEnabledWithStatusBar
-           || _shiftBehavior == MDCFlexibleHeaderShiftBehaviorHideable) &&
+          (_shiftBehavior == MDCFlexibleHeaderShiftBehaviorEnabledWithStatusBar ||
+           _shiftBehavior == MDCFlexibleHeaderShiftBehaviorHideable) &&
```

<!-- identifier: clang-format -->
<!-- File: components/FlexibleHeader/src/MDCFlexibleHeaderView.m -->

material-automation commented on components/FlexibleHeader/src/MDCFlexibleHeaderView.m:
```diff
@@ -1669,7 +1674,8 @@ - (BOOL)prefersStatusBarHidden {
 - (BOOL)hidesStatusBarWhenCollapsed {
-  return (_shiftBehavior == MDCFlexibleHeaderShiftBehaviorEnabledWithStatusBar &&
+  return ((_shiftBehavior == MDCFlexibleHeaderShiftBehaviorEnabledWithStatusBar
+           || _shiftBehavior == MDCFlexibleHeaderShiftBehaviorHideable) &&
```
clang-format suggested the following change:


```
-  return ((_shiftBehavior == MDCFlexibleHeaderShiftBehaviorEnabledWithStatusBar
-           || _shiftBehavior == MDCFlexibleHeaderShiftBehaviorHideable) &&
+  return ((_shiftBehavior == MDCFlexibleHeaderShiftBehaviorEnabledWithStatusBar ||
+           _shiftBehavior == MDCFlexibleHeaderShiftBehaviorHideable) &&
```

<!-- identifier: clang-format -->
<!-- File: components/FlexibleHeader/src/MDCFlexibleHeaderView.m -->

jverkoey commented:
First PR out at https://github.com/material-components/material-components-ios/pull/9014

Commit:
[AppBar] Add a shouldSetNavigationBarHiddenHideAppBar behavior flag.

This new flag allows view controllers to control the visibility of their app bar via the standard UINavigationController setNavigationBarHidden: APIs. When enabled, calls to these APIs will result in the AppBar's visibility being adjusted.

Closes https://github.com/material-components/material-components-ios/issues/5185

jverkoey commented:
Tests are outstanding still. I will try to get them as far along as possible before I sign off today.

yarneo commented on components/AppBar/src/MDCAppBarNavigationController.m:
```diff
@@ -64,8 +64,9 @@ - (instancetype)initWithRootViewController:(UIViewController *)rootViewControlle
-  [self setNavigationBarHidden:YES animated:NO];
+  // We always want the UIKit navigation bar to be hidden; to do so we must invoke the super
+  // implementation.
+  [super setNavigationBarHidden:YES animated:NO];
```
nit: you can call the API without the animated param if you don't wish to animate.

jverkoey replied:
I use the `animate:` variant here mainly to be explicit.

yarneo replied:
ack

yarneo commented on components/AppBar/src/MDCAppBarNavigationController.m:
```diff
@@ -102,23 +103,78 @@ - (void)setViewControllers:(NSArray<UIViewController *> *)viewControllers animat
+  }
+
+  if (appBarViewController.headerView.shiftBehavior == MDCFlexibleHeaderShiftBehaviorDisabled) {
+    appBarViewController.headerView.shiftBehavior = MDCFlexibleHeaderShiftBehaviorHideable;
```
doesn't this depend on the value of navigationBarHidden?

jverkoey replied:
In what way do you see it depending on the value? Please expand on what you see that may be missing here.

yarneo replied:
I don't fully understand why the shiftBehavior is changing here regardless of what value the setter received, hence me making sure this wasn't missed. If this isn't affected by the setter value, then does this block of code need to be part of the setter?

jverkoey replied:
> If this isn't affected by the setter value

Can you clarify what "this" and "the setter value" means in this context?

Does "this" refer to this line of assignment?

And does "the setter value" refer to the fact that we're assigning `MDCFlexibleHeaderShiftBehaviorHideable` here?

jverkoey replied:
While waiting for clarification, the English version of this code is as follows:

"If the shift behavior is presently disabled (the default), then adjust it to be hideable instead, otherwise we do not make any adjustments to the shift behavior because it's likely that the shift behavior was explicitly set to something else."

jverkoey replied:
Added this explanation as docs to the code.

yarneo replied:
this = appBarViewController.headerView.shiftBehavior = MDCFlexibleHeaderShiftBehaviorHideable;

setter value = navigationBarHidden

jverkoey replied:
Ah I see now. Yes we're changing the behavior of the app bar the moment these APIs are used (regardless of the visibility) so that the behavior is instantly "locked in" to the correct `hideable` behavior of the flexible header.

Commit:
Rearrange.

yarneo commented on components/AppBar/src/MDCAppBarNavigationController.m:
```diff
@@ -102,23 +103,78 @@ - (void)setViewControllers:(NSArray<UIViewController *> *)viewControllers animat
+  }
+
+  // Only toggle visiblity on app bars that have the correct shift behavior enabled.
+  if (appBarViewController.headerView.shiftBehavior != MDCFlexibleHeaderShiftBehaviorHideable) {
```
```
if (appBarViewController.headerView.shiftBehavior == MDCFlexibleHeaderShiftBehaviorDisabled) {
  appBarViewController.headerView.shiftBehavior = MDCFlexibleHeaderShiftBehaviorHideable;
} else if (appBarViewController.headerView.shiftBehavior != MDCFlexibleHeaderShiftBehaviorHideable) {
  return;
}
```

jverkoey replied:
I'm intentionally funneling all logical paths above through to this final gate in order to reduce the number of logical paths for completion of this method. The check here is acting in this regard as a `guard` statement.

yarneo replied:
Yes I know, but the above snippet provides the exact same logic with needing less cycles if the shiftBehavior was Disabled. See it as a performance tip, no action needed.

jverkoey replied:
This method is not going to be on any critical path so I'd prefer to optimize for readability than performance here.

yarneo replied:
ack

yarneo reviewed (approved):
LGTM % tests

Commit:
Basic test.

Commit:
More basic tests.

Commit:
Docs.

Commit:
More tests.

Commit:
More tests.

Commit:
More test.

Commit:
Two clangs diverged in a yellow wood,

jverkoey commented:
Tests are done.

material-automation commented on components/AppBar/tests/unit/AppBarNavigationControllerNavigationBarHiddenTests.m:
```diff
@@ -0,0 +1,57 @@
+- (void)testStillHiddenAfterSettingNavigationBarHiddenAnimated {
+  // Given
+  MDCAppBarNavigationController *appBarNavigationController =
+  [[MDCAppBarNavigationController alloc] init];
```
clang-format suggested the following change:


```
-  [[MDCAppBarNavigationController alloc] init];
+      [[MDCAppBarNavigationController alloc] init];
```

<!-- identifier: clang-format -->
<!-- File: components/AppBar/tests/unit/AppBarNavigationControllerNavigationBarHiddenTests.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

material-automation commented on components/AppBar/tests/unit/AppBarNavigationControllerNavigationBarHiddenWhenNavigationBarHiddenHidesAppBarTests.m:
```diff
@@ -0,0 +1,53 @@
+
+#import "MaterialAppBar.h"
+
+@interface AppBarNavigationControllerNavigationBarHiddenWhenNavigationBarHiddenHidesAppBarTests : XCTestCase
```
clang-format suggested the following change:


```
-@interface AppBarNavigationControllerNavigationBarHiddenWhenNavigationBarHiddenHidesAppBarTests : XCTestCase
+@interface AppBarNavigationControllerNavigationBarHiddenWhenNavigationBarHiddenHidesAppBarTests
+    : XCTestCase
```

<!-- identifier: clang-format -->
<!-- File: components/AppBar/tests/unit/AppBarNavigationControllerNavigationBarHiddenWhenNavigationBarHiddenHidesAppBarTests.m -->
